benchmarks-mobile/protonano_app/app/src/main/java/io/grpc/benchmarks/ProtobufBenchmarker.java

```diff
@@ -0,0 +1,165 @@
+                    new Action() {
+                @Override
+                public void execute() throws IOException {
+                    ByteArrayOutputStream bos = new ByteArrayOutputStream(serializedSize);
+                    GZIPOutputStream gos = new GZIPOutputStream(bos);
+                    gos.write(jsonObject.toString().getBytes());
+                    gos.close();
+                    bos.close();
+                    bos.toByteArray();
+                }
+            });
+            res.compressedSize = bos.toByteArray().length;
+            return res;
+        } else {
+            return benchmark("JSON serialize to byte array", serializedSize, new Action() {
+                @Override
+                public void execute() throws JSONException {
+                    jsonObject.toString().getBytes();
+                }
+            });
+        }
+    }
+
+    public static BenchmarkResult deserializeJsonfromByteArray(final String jsonString,
+                                                               boolean gzip) throws Exception {
+        final int serializedSize = jsonString.getBytes().length;
+
+        if (gzip) {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream(serializedSize);
+            GZIPOutputStream gos = new GZIPOutputStream(bos);
+            gos.write(jsonString.getBytes());
+            gos.close();
+            bos.close();
+            final byte[] compressedData = bos.toByteArray();
+
+            BenchmarkResult res = benchmark("JSON deserialize from byte array (gzip)",
+                    serializedSize, new Action() {
+                @Override
+                public void execute() throws JSONException, IOException {
+                    ByteArrayInputStream bis = new ByteArrayInputStream(compressedData);
+                    GZIPInputStream gis = new GZIPInputStream(bis, serializedSize);
+                    byte[] inputData = new byte[serializedSize];
+                    gis.read(inputData);
+                    gis.close();
+                    bis.close();
+                    new JSONObject(new String(inputData));
+                }
+            });
+            res.compressedSize = compressedData.length;
+            return res;
+        } else {
+            final byte[] jsonData = jsonString.getBytes();
+            return benchmark("JSON deserialize from byte array", serializedSize, new Action() {
+                @Override
+                public void execute() throws JSONException {
+                    new JSONObject(new String(jsonData));
+                }
+            });
+        }
+    }
+
+    private static BenchmarkResult benchmark(String name, long dataSize, Action action) throws Exception {
+        for (int i = 0; i < 100; ++i) {
+            action.execute();
+        }
+
+        int iterations = 1;
+        long elapsed = timeAction(action, iterations);
+        while (elapsed < MIN_SAMPLE_TIME_MS) {
```

Comment:
This is Varying in the wrong direction.  You should instead hold the runtime constant, and record the number of iterations.  Start another thread, and then make this one go to sleep for 1 second.  When it wakes up it should count the number of iterations.  

This is the approach used by real benchmarking frameworks.